File: atob.c
/*
 * atob: Convert ascii text from standard input to readable binary.
 * Written by William S. Moore
 */
#include <stdio.h>

int main(int argc, char *argv[])
{
	if (argc > 1) {
		printf("atob: Convert ascii text from standard input to readable binary.\n");
		printf("Note: The newline character '\\n' is presereved in the output.\n");
		printf("You are reading this because arguments were passed to atob.\n");
		return 1;
	}

	int c;
	int first_in_line = 1; // flag to track the start of a line
	int space_offset = 0;
	char buffer[10]; // 8 bits + possible space + null terminator

	while((c = getchar()) != EOF) {
		if (c == '\r') continue; // discard carrige returns

		if (!first_in_line) {
			buffer[0] = ' ';
			space_offset = 1;
		}

		unsigned char mask = 0x80; // start at bit 8 (index 7)
		for (int i = 0 + space_offset; i < 8 + space_offset; i++) {
			buffer[i] = (c & mask) ? '1' : '0';
			mask >>= 1;
		}
		buffer[8 + space_offset] = '\0';
		space_offset = 0;

		fputs(buffer, stdout);

		if (c == '\n') {
			putchar('\n');
			first_in_line = 1;
		} else {
			first_in_line = 0;
		}
	}
	return 0;
}

File: AtoB.java
import java.io.IOException;

/**
 * atob: Convert ASCII text from standard input to readable binary.
 * Ported from C by Gemini.
 */
public class AtoB {
    public static void main(String[] args) {
        // C logic: if (argc > 1)
        if (args.length > 0) {
            System.out.println("atob: Convert ascii text from standard input to readable binary.");
            System.out.println("Note: The newline character '\\n' is preserved in the output.");
            System.out.println("You are reading this because arguments were passed to atob.");
            System.exit(1);
        }

        try {
            int c;
            boolean firstInLine = true;

            while ((c = System.in.read()) != -1) {
                if (c == '\n') {
                    System.out.print('\n');
                    firstInLine = true;
                } 
                else if (c == '\r') {
                    // Skip carriage returns to match C behavior
                    continue;
                } 
                else {
                    // Add a space between binary blocks, but not at the very start of a line
                    if (!firstInLine) {
                        System.out.print(' ');
                    }

                    // Bit manipulation logic
                    int mask = 0x80; // 10000000 in binary
                    StringBuilder buffer = new StringBuilder();
                    
                    for (int i = 0; i < 8; i++) {
                        buffer.append((c & mask) != 0 ? '1' : '0');
                        mask >>= 1;
                    }
                    
                    System.out.print(buffer.toString());
                    firstInLine = false;
                }
            }
        } catch (IOException e) {
            System.err.println("Error reading input: " + e.getMessage());
        }
    }
}
